### PR TITLE
AE-961: Event Monitor condition check and tests

### PR DESCRIPTION
**Pull request by arbitrary-sparrow**
This PR adds an optional condition check to the admin events agent template and adds tests. The following sample `agent-config.json` can be used to test this logic:
```json
{
  "developerAbbreviation": "AE",
  "protocolName": "UNISWAP",
  "protocolAbbreviation": "UNI",
  "contracts": {
    "GovernorBravo": {
      "address": "0x408ED6354d4973f66138C91495F2f2FCbd8724C3",
      "abiFile": "GovernorBravo.json",
      "events": {
        "ProposalCreated": {
          "expression": "proposer !== 0x9B68c14e936104e9a7a24c712BEecdc220002984",
          "type": "Info",
          "severity": "Info"
        }
      }
    }
  }
}
```

Additionally, a simplified ABI for `GovernorBravo.json` is as follows:
```json
{
  "abi":
  [
    {
      "anonymous": false,
      "inputs": [
        {
          "indexed": false,
          "internalType": "uint256",
          "name": "id",
          "type": "uint256"
        },
        {
          "indexed": false,
          "internalType": "address",
          "name": "proposer",
          "type": "address"
        },
        {
          "indexed": false,
          "internalType": "address[]",
          "name": "targets",
          "type": "address[]"
        },
        {
          "indexed": false,
          "internalType": "uint256[]",
          "name": "values",
          "type": "uint256[]"
        },
        {
          "indexed": false,
          "internalType": "string[]",
          "name": "signatures",
          "type": "string[]"
        },
        {
          "indexed": false,
          "internalType": "bytes[]",
          "name": "calldatas",
          "type": "bytes[]"
        },
        {
          "indexed": false,
          "internalType": "uint256",
          "name": "startBlock",
          "type": "uint256"
        },
        {
          "indexed": false,
          "internalType": "uint256",
          "name": "endBlock",
          "type": "uint256"
        },
        {
          "indexed": false,
          "internalType": "string",
          "name": "description",
          "type": "string"
        }
      ],
      "name": "ProposalCreated",
      "type": "event"
    }
  ]
}
```

### CODE REVIEW

File: admin-events/SETUP.md [hunk 1]
```diff
@@ -10,52 +10,76 @@ The following steps will take you from a completely blank template to a function
...
-5. `protocolAbbreviation` (required) - Type in an appropriate abbreviation for the value in `protocolName`.  For
+4. `protocolAbbreviation` (required) - Type in an appropriate abbreviation for the value in `protocolName`.  For
 example, `"Uniswap"` may be abbreviated `"UNI"` and `"SushiSwap"` may be abbreviated `"SUSH"`, etc.
 
-6.  `adminEvents` (required) - The Object value for this key corresponds to contracts that we want to monitor events
+5.  `contracts` (required) - The Object value for this key corresponds to contracts that we want to monitor events
 for.  Each key in the Object is a contract name that we can specify, where that name is simply a string that we use
 as a label when referring to the contract (the string can be any valid string that we choose, it will not affect the
 monitoring by the agent).  The Object corresponding to each contract name requires an address key/value pair, abi
 file key/value pair, and either an `events` key or a `proxy` key.  For the case of an `events` key, the corresponding
```
Comment (echoAE) [inline admin-events/SETUP.md:26]:
```suggestion
file key/value pair, and an `events` key, `proxy` key, or both.  For the case of an `events` key, the corresponding
```

File: admin-events/SETUP.md [hunk 2]
```diff
@@ -173,14 +203,25 @@ reflect who is creating the agent and what the agent monitors.
...
-11. Install all related `npm` packages using `npm i`.  This will create a `package-lock.json` file alongside
+10. Install all related `npm` packages using `npm i`.  This will create a `package-lock.json` file alongside
 package.json.
 
-12. Once the `agent-config.json` file is populated and all corresponding ABI files are in the correct locations
+11. Once the `agent-config.json` file is populated and all corresponding ABI files are in the correct locations
 referred to in the `agent-config.json` file, the agent is complete.  Please test the agent against transactions
 that contain events that should trigger the agent.  Please also test the agent against transactions that should
 not trigger the agent.  Although not provided here, please create tests that will verify the functionality of
 the agent code for positive cases, negative cases, and edge cases (e.g. when errors occur).
```
Comment (echoAE) [inline admin-events/SETUP.md:213]:
```suggestion
not trigger the agent.
```

File: admin-events/src/utils.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    case '>=':
+      return variable.gte(operand);
+    case '>':
+      return variable.gt(operand);
+    case '<=':
+      return variable.lte(operand);
+    case '<':
+      return variable.lt(operand);
+    default:
+      throw new Error(`BigNumber operator ${operator} no supported`);
```
Comment (echoAE) [inline admin-events/src/utils.js:85]:
```suggestion
      throw new Error(`BigNumber operator ${operator} not supported`);
```

File: admin-events/src/utils.spec.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      const { comparisonFunction } = parseExpression('value === true');
+      expect(comparisonFunction).toBe(booleanComparison);
+    });
+
+    it('returns an object containing the bigNumberComparison function if the expression supplied is a number condition', () => {
+      const { comparisonFunction } = parseExpression('x >= 123456');
+      expect(comparisonFunction).toBe(bigNumberComparison);
+    });
+
+    it('throws an error if an invalid operand is passed for an address condition', () => {
```
Comment (echoAE) [inline admin-events/src/utils.spec.js:114]:
```suggestion
    it('throws an error if an invalid operator is passed for an address condition', () => {
```

File: admin-events/src/utils.spec.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+      expect(() => {
+        parseExpression('address >= 0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
+      }).toThrow('Unsupported address operator ">=": must be "===" or "!=="');
+    });
+
+    it('throws an error if an invalid operand is passed for a boolean condition', () => {
```
Comment (echoAE) [inline admin-events/src/utils.spec.js:120]:
```suggestion
    it('throws an error if an invalid operator is passed for a boolean condition', () => {
```

File: admin-events/src/utils.spec.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      expect(() => {
+        parseExpression('true > false');
+      }).toThrow('Unsupported Boolean operator ">": must be "===" or "!=="');
+    });
+
+    it('throws an error if an invalid operand is passed for a number condition', () => {
```
Comment (echoAE) [inline admin-events/src/utils.spec.js:126]:
```suggestion
    it('throws an error if an invalid operator is passed for a number condition', () => {
```

File: admin-events/src/test-utils.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+          leftOperand = value.plus(1).toString();
+        } else {
+          leftOperand = value.toString();
+        }
+        break;
+      case '<':
+        if (expectedResult) {
+          leftOperand = value.minus(1).toString();
+        } else {
+          leftOperand = value.plus(1).toString();
```
Comment (echoAE) [inline admin-events/src/test-utils.js:96]: Ubernit: In all of the other cases, we either set `leftOperand` equal to `value` or we intentionally set it +-1.  In this case, we broke from that pattern.

File: admin-events/src/test-utils.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+      case '!==':
+        if (expectedResult) {
+          leftOperand = ethers.constants.AddressZero;
+        } else {
+          leftOperand = value;
+        }
+        break;
+      default:
+        throw new Error(`Unsupported operator ${operator} for hexString comparison`);
+    }
```
Comment (echoAE) [inline admin-events/src/test-utils.js:120]: The method `ethers.utils.isHexString` will check that the argument is of type `string` and starts with `0x` and contains only characters `0-9A-Fa-f` afterward.  Because an address string meets these qualifications, the `else if` block after this will never be run.

It probably makes sense to completely deprecate the `isAddress` function in favor of the `ethers` method.  There is an optional second argument to `ethers.utils.isHexString` to specify the length of the hex string, allowing us to achieve the same functionality as `isAddress`.

Also, lines 108 and 113 will work for every case except where `value` is equal to `ethers.constants.AddressZero`.

Comment (arbitrary-sparrow) [reply admin-events/src/test-utils.js]: I ended up wrapping a `ether.utils.isHexString(..., 20)` call in the `isAddress` function

File: admin-events/src/agent.spec.js [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    });
+
+    it('returns empty findings if contract address matches but no monitored function was invoked', async () => {
+      // encode event data - valid event with valid arguments
+      const { mockArgs, mockTopics, data } = createMockEventLogs(eventNotInConfig, iface);
+
+      // update mock transaction event
+      const [defaultLog] = mockTxEvent.receipt.logs;
+      defaultLog.name = contractName;
+      defaultLog.address = ethers.constants.AddressZero;
```
Comment (echoAE) [inline admin-events/src/agent.spec.js:208]:
```suggestion
      defaultLog.address = validContractAddress;
```

File: admin-events/src/agent.spec.js [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+          eventName: eventInConfig.name,
+          ...expectedMetaData,
+        },
+      })];
+
+      expect(findings).toStrictEqual(testFindings);
+    });
+
+    it('returns a finding if a target contract emits a monitored event and the expression condition is met', async () => {
+      // get the expression object information from the config
```
Comment (echoAE) [inline admin-events/src/agent.spec.js:271]:
```suggestion
      // get the expression object information from the config
      // in the beforeEach block, the first event from the first `contracts` element is assigned to `eventInConfig`
      // therefore, we will retrieve the corresponding expression from the `initializeData` object to enforce the
      // proper condition for this test to emit a finding
```

**Review by echoAE — COMMENTED** [review]
A few changes suggested.

**Review by echoAE — APPROVED** [review]

File: admin-events/SETUP.md [hunk 11]
```diff
@@ -10,52 +10,76 @@ The following steps will take you from a completely blank template to a function
...
 entries for running the agent.
 
-7. If a contract is a proxy for another contract, where events will be emitted as if they are coming from the proxy
+6. If a contract is a proxy for another contract, where events will be emitted as if they are coming from the proxy
 instead of from the underlying implementation contract, the entry in the `agent-config.json` file may look like the
 following:
 
-```
+```json
   "adminEvents": {
```
Comment (arbitrary-CodeBeholder) [inline admin-events/SETUP.md:83]: this should be `"contracts"`

File: admin-events/agent-config.json [hunk 12]
```diff
@@ -1,21 +1,27 @@
 {
```
Comment (arbitrary-CodeBeholder) [inline admin-events/agent-config.json:1]: it appears that this config actually contains an example config, so the filename should probably be changed to something like `agent-config.json.example` that way it can be copied to the use in the actual config and users won't think it's a real config file

**Review by arbitrary-CodeBeholder — APPROVED** [review]
Most of the same nits as the #8
